src/rust/ide/lib/view/src/graph_editor/component/breadcrumbs.rs

```diff
@@ -0,0 +1,337 @@
+        let scene                 = scene.clone_ref();
+        let breadcrumbs           = default();
+        let breadcrumb_pop        = frp.outputs.breadcrumb_pop.clone_ref();
+        let breadcrumb_push       = frp.outputs.breadcrumb_push.clone_ref();
+        let current_index         = default();
+        Self{logger,display_object,animations,scene,breadcrumbs,breadcrumb_pop,breadcrumb_push,
+            project_name,breadcrumbs_container,current_index}.init()
+    }
+
+    fn init(self) -> Self {
+        self.add_child(&self.project_name);
+        self.add_child(&self.breadcrumbs_container);
+        self.project_name.set_position(Vector3::new(HORIZONTAL_MARGIN,0.0,0.0));
+        self.set_project_name_width(self.project_name.width());
+        //FIXME[dg]: This is a workaround. Main shouldn't be a MethodPointer.
+        let mut method_pointer = enso_protocol::language_server::MethodPointer::default();
+        method_pointer.name    = "Main".to_string();
+        let method_pointer     = Rc::new(method_pointer);
+        self.push_breadcrumb(&method_pointer,&default());
+        self
+    }
+
+    fn width(&self) -> f32 {
+        self.breadcrumbs.borrow().iter().fold(0.0, |acc,breadcrumb| acc + breadcrumb.width())
+    }
+
+    fn set_project_name_width(&self, width:f32) {
+        self.breadcrumbs_container.set_position(Vector3::new(HORIZONTAL_MARGIN + width,0.0,0.0));
+    }
+
+    fn select_breadcrumb(&self, index:usize) {
+        let current_index = self.current_index.get();
+        match index.cmp(&current_index) {
+            cmp::Ordering::Less => {
+                // If we have more crumbs after `index`, we will pop them.
+                let pop_amount    = current_index - index;
+                for _ in 0..pop_amount {
+                    self.breadcrumb_pop.emit(());
+                }
+            },
+            cmp::Ordering::Greater => {
+                for index in current_index..index {
+                    let info = self.breadcrumbs.borrow().get(index).map(|breadcrumb| {
+                        (breadcrumb.info.method_pointer.clone(),breadcrumb.info.expression_id)
+                    }).as_ref().cloned();
+                    if let Some((method_pointer,expression_id)) = info {
+                        self.breadcrumb_push.emit((method_pointer,expression_id));
+                    }
+                }
+            },
+            cmp::Ordering::Equal => ()
+        }
+        info!(self.logger,"Selecting breadcrumb #{index}");
+    }
+
+    fn push_breadcrumb(&self, method_pointer:&Rc<MethodPointer>, expression_id:&uuid::Uuid) {
+        let current_index = self.current_index.get();
+        let next_index    = current_index + 1;
+
+        let breadcrumb_exists =
+            self.breadcrumbs.borrow_mut().get(current_index).map(|breadcrumb| {
+                breadcrumb.info.expression_id
+            }).filter(|other_expression_id| other_expression_id == expression_id).is_some();
```

Comment:
i think this map + filter + is_sone can be simplified to our `contains_if`

Reply:
move it to `OptionOps` in the prelude, I like it too much to leave it limited to a `ide` crate